docs: Implemented "recipe" page

aminaopio opened the pull request:
## What changed / motivation ?

Implemented a recipe page within the docs that contains a bunch of common styling patterns.

## Additional Context

<!--- Screenshots, Tests, Breaking Change, Anything Else ? --->
<img width="500" alt="Screenshot 2024-12-04 at 5 28 41 PM" src="https://github.com/user-attachments/assets/c8ff75d1-853a-4173-95ea-c93021fd737b">
<img width="500" alt="Screenshot 2024-12-04 at 5 28 48 PM" src="https://github.com/user-attachments/assets/c0cd0f77-7e7c-478b-b4c0-8bd2a22b83fd">
<img width="500" alt="Screenshot 2024-12-04 at 5 28 57 PM" src="https://github.com/user-attachments/assets/8bc6e426-f126-4db6-8280-1ccb7392d3aa">



## Pre-flight checklist

- [x] I have read the contributing guidelines
      [Contribution Guidelines](https://github.com/facebook/stylex/blob/main/.github/CONTRIBUTING.md)
- [x] Performed a self-review of my code

p0nch000 commented:
if there's something else to add to the docs, I can help with it lmk

nmn commented:
Can we break up the recipes into separate pages so we can put more detail into each recipe over time?

nmn commented on apps/docs/docs/learn/04-styling-ui/03-recipes.mdx:
````diff
@@ -0,0 +1,152 @@
+```tsx
+import * as stylex from '@stylexjs/stylex';
+
+const styles = stylex.create({
+  primary: {
+    backgroundColor: 'blue',
+    color: 'white',
+    ':hover': {
+      backgroundColor: 'darkblue',
+    },
+  },
+  secondary: {
+    backgroundColor: 'gray',
+    color: 'white',
+    ':hover': {
+      backgroundColor: 'darkgray',
+    },
+  },
+});
````
Don't use the deprecated and incorrect way to use `:hover`

```suggestion
const styles = stylex.create({
  primary: {
    backgroundColor: {
      default: 'blue',
      ':hover': 'darkblue',
    },
    color: 'white',
  },
  secondary: {
    backgroundColor: {
      default: 'gray',
      ':hover': 'darkgray',
    },
    color: 'white',
  },
});
```

nmn commented on apps/docs/docs/learn/04-styling-ui/03-recipes.mdx:
```diff
@@ -0,0 +1,152 @@
+});
+
+function Button({ variant = 'primary', ...props }) {
+  return <button {...props} {...stylex.props(styles[variant])} />;
```
```suggestion
  return <button {...props} {...stylex.props(styles[variant], props.style)} />;
```

nmn commented on apps/docs/docs/learn/04-styling-ui/03-recipes.mdx:
````diff
@@ -0,0 +1,152 @@
+```tsx
+import * as stylex from '@stylexjs/stylex';
+
+const styles = stylex.create({
+  parent: {
+    '--child-color': 'black',
+    ':hover': {
+      '--child-color': 'blue',
+    },
+  },
+  child: {
+    color: 'var(--child-color)',
+  },
+});
````
Change this example to use a separate `.stylex.ts` file that defines a variable using `stylex.defineVars`. We discourage the use of arbitrary strings for variables.

nmn commented on apps/docs/docs/learn/04-styling-ui/03-recipes.mdx:
```diff
@@ -0,0 +1,152 @@
+
+The best approach to building responsive layouts is often to use fluid designs that adapt naturally to available space. This minimizes the total CSS and avoids brittle breakpoints.
+
+### Example: Fluid Grid Layout
```
A fluid grid should probably be implemented with `display: grid` and not flexbox.

nmn replied:
Just remove this example for now. This is styling advice and not a StyleX recipe.

aminaopio commented:
> Can we break up the recipes into separate pages so we can put more detail into each recipe over time?

Gotcha, I implemented your suggestions, and added a folder for recipes with each individual page separated. Should be good to go, but let me know

nmn commented on apps/docs/docs/learn/06-recipes/01-contexual-styles.mdx:
```diff
@@ -0,0 +1,45 @@
+
+const SidebarContext = createContext(false);
+
+const styles = stylex.create({
+  sidebarOpen: {
+    width: '250px',
+  },
+  sidebarClosed: {
+    width: '50px',
+  },
+});
```
```suggestion
const styles = stylex.create({
  sidebarOpen: {
    width: 250,
  },
  sidebarClosed: {
    width: 50,
  },
});
```

nmn commented on apps/docs/docs/learn/06-recipes/01-contexual-styles.mdx:
```diff
@@ -0,0 +1,45 @@
+import { createContext, useContext } from 'react';
+import * as stylex from '@stylexjs/stylex';
+
+const SidebarContext = createContext(false);
```
You might want to show three files for this example and show how a parent component can use `SidebarContext.Provider` to provide the value down.

nmn commented on apps/docs/docs/learn/06-recipes/02-variants.mdx:
````diff
@@ -0,0 +1,47 @@
+```tsx
+import * as stylex from '@stylexjs/stylex';
+
+const styles = stylex.create({
+  primary: {
+    backgroundColor: {
+      default: 'blue',
+      ':hover': 'darkblue',
+    },
+    color: 'white',
+  },
+  secondary: {
+    backgroundColor: {
+      default: 'gray',
+      ':hover': 'darkgray',
+    },
+    color: 'white',
+  },
+});
+
+function Button({ variant = 'primary', ...props }) {
+  return <button {...props} {...stylex.props(styles[variant], props.style)} />;
+}
````
It might make sense to see the examples used by [CVA](https://cva.style/docs/getting-started/variants) and provide equivalent examples. Therefore, it makes sense to show "base" styles that stay the same for all variants and the variants themselves.

```suggestion
const styles = stylex.create({
  base: {
    appearance: 'none',
    borderWidth: 0,
  },
});

const variants = stylex.create({
  primary: {
    backgroundColor: {
      default: 'blue',
      ':hover': 'darkblue',
    },
    color: 'white',
  },
  secondary: {
    backgroundColor: {
      default: 'gray',
      ':hover': 'darkgray',
    },
    color: 'white',
  },
});

function Button({ variant = 'primary', ...props }) {
  return <button {...props} {...stylex.props(styles.base, variants[variant], props.style)} />;
}
```

nmn commented on apps/docs/docs/learn/06-recipes/03-descendant styles.mdx:
```diff
@@ -0,0 +1,46 @@
+    },
+  },
+  child: {
+    color: `var(${vars.childColor})`,
```
You don't have to wrap a variable created with StyleX with `var()`. It works as a reference as is.

```suggestion
    color: vars.childColor,
```

nmn commented on apps/docs/docs/learn/06-recipes/01-contexual-styles.mdx:
```diff
@@ -0,0 +1,45 @@
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+sidebar_position: 5
```
You need to update these numbers to order the recipes. I suggest putting `Variants` first as that is a common use-case.

nmn reviewed (changes requested):
A few more changes needed, but nearly there.

aminaopio commented:
> A few more changes needed, but nearly there.

Think it's done

nmn commented on apps/docs/docs/learn/06-recipes/01-contexual-styles.mdx:
```diff
@@ -0,0 +1,94 @@
+const SidebarContext = createContext(false);
+
+const styles = stylex.create({
+  sidebarOpen: {
+    width: '250',
+  },
+  sidebarClosed: {
+    width: '50',
+  },
```
The suggested code did not have quotes around the numbers.

nmn reviewed (commented):
Still a few issues. I'll fix them and merge it.

aminaopio commented:
@nmn Forgot about this one, but let me know if anything else can be done on my end!

nmn commented:
@aminaopio This is good for now. I will merge it soon.
